### PR TITLE
Extract selection strategies to classes

### LINKED ISSUE #122: De-duplicate Dynamic[Cluster/Thread]Pool chooseWorker with an implementation of the strategy design pattern
**Is your feature request related to a problem? Please describe.**
Currently we have many LOC duplicated in these both `chooseWorker` functions

**Describe the solution you'd like**
Find a strategy to de duplicate this part of code and apply it

**Additional context**
https://github.com/pioardi/poolifier/pull/116#discussion_r574106019  
https://github.com/pioardi/poolifier/pull/116#issuecomment-777469240

---

A first try:

```ts
type IDynamicPool<Worker extends IWorker, Data, Response> = (
  | DynamicClusterPool<Data, Response>
  | DynamicThreadPool<Data, Response>
) &
  AbstractPool<Worker, Data, Response>

export function dynamicWorkerChooser<
  Worker extends IWorker,
  Data,
  Response,
  DynamicPool extends IDynamicPool<Worker, Data, Response>
> (this: DynamicPool, superChooseWorker: () => Worker): Worker {
  // ...
```

And then it could be called e.g. like this:
```ts
return dynamicWorkerChooser.bind(this, () => super.chooseWorker())
```

But when accessing functions like `internalNewWorker` or `sendToWorker` in `this(: DynamicPool)`, it cannot access them, because it tells me that `this` has no method defined like the ones :confused:

### PR DESCRIPTION
**Pull request by Shinigami92**
## PR Checklist

- [x] Please add a description of your changes.
- [ ] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [ ] Please add a link to the open issue or task that this pull request will resolve.
- [x] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

closes #122

### CODE REVIEW

**Comment by pioardi** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=pioardi_poolifier&pullRequest=176&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=pioardi_poolifier&pullRequest=176&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=pioardi_poolifier&pullRequest=176&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=pioardi_poolifier&pullRequest=176&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=pioardi_poolifier&pullRequest=176&metric=new_coverage&view=list) [100.0% Coverage](https://sonarcloud.io/component_measures?id=pioardi_poolifier&pullRequest=176&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=pioardi_poolifier&pullRequest=176&metric=new_duplicated_lines_density&view=list) [21.4% Duplication](https://sonarcloud.io/component_measures?id=pioardi_poolifier&pullRequest=176&metric=new_duplicated_lines_density&view=list)

@Shinigami92  could you please check the code duplicatio?

**Comment by Shinigami92** [comment]
image

**Comment by pioardi** [comment]
@Shinigami92 great so we Will soon made the sonar task mandatory to merge a PR.
General shape Is good I Will review Better tomorrow ( from phone currently ).
I was thinking : should  we emit the fullpool event also for fixed pools? This Is not directly related to your changes

**Comment by Shinigami92** [comment]
> I was thinking : should we emit the fullpool event also for fixed pools? This Is not directly related to your changes

But for a fixed pool it would mean something else
dynamic FullPool => no worker could be chosen, cause **max** was reached
fixed ??? => all workers are busy?

File: src/pools/thread/dynamic.ts
Comment (pioardi) [inline src/pools/thread/dynamic.ts:57]: could we save the binded functions and reuse them instead of do the bind each time ? 
We could also try to draft a class instead of a single method so we can store those binded functions into the class instead of pass them each time ( performance reason ).

Please let me know your thoughts

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [inline src/pools/selection-strategies.ts:10]: You want to define a type for each selection algo?

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [inline src/pools/selection-strategies.ts:71]: That one shall be a class. 
The strategy design pattern use class to define strategy for registration reason: set then execute. 
With functions the call stack is really deep and have impact on performance.

Comment (pioardi) [reply src/pools/selection-strategies.ts:71]: Yes I also made a similar comment on the caller side

File: src/pools/thread/dynamic.ts
Comment (Shinigami92) [reply src/pools/thread/dynamic.ts:57]: The current problem is, that most of these functions are protected by design.

Are there performance reasons?

So please provide / draft an alternative solution if you have one

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [reply src/pools/selection-strategies.ts:10]: No, this was just for easier JSDoc and typedef
It is also not exported via `index.ts` by design

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [reply src/pools/selection-strategies.ts:71]: Interesting, but how can I (you someone of you) do that using protected or private member functions?
The current solution preserves the member function's accessors. That's why it's build so complicated.
If we just made all of these functions `public` it would be much easier, but much more dangerous because users can do everything they want (what they already can because accessors are just TS, but they shall not...)

Comment (pioardi) [reply src/pools/selection-strategies.ts:71]: Functions can just be public must not exported to the end users I guess

File: src/pools/thread/dynamic.ts
Comment (pioardi) [reply src/pools/thread/dynamic.ts:57]: @Shinigami92  not sure I got your point could you please elaborate more on that?
We can just set the functions as protected you know, like we do with others

Yes there are performance reasons using the bind function each time will decrease the performance.

Comment (Shinigami92) [reply src/pools/thread/dynamic.ts:57]: Maybe I'm overlooking something, but when I pass e.g. the instance like `dynamicallyChooseWorker(this)` I don't have access to e.g. `that.createAndSetupWorker()` because `createAndSetupWorker` is a protected member function
So how should I pass these functions to `dynamicallyChooseWorker` so I can successfully call them within `dynamicallyChooseWorker`?

---

Feel free to create a PR stacked on this PR and show some experiments :slightly_smiling_face:

Comment (pioardi) [reply src/pools/thread/dynamic.ts:57]: `
// do this in some point where the code is executed one time ( maybe the constructor could be fine)
const myBindedFunction = this.createAndSetupWorker.bind(this)  

// or
this.myBindedFunction = this.createAndSetupWorker.bind(this) // in this case we could need to set it as protected in some way
`

and  


`
dynamicallyChooseWorker(..,..,..,myBindedFunction)
`

**Comment by Shinigami92** [comment]
@pioardi Please make a new review
I also added `IDynamicPool`

I'm personally against the `isDynamic`, because maybe we want to add more and more pools and I don't want to have an "everything" pool but via settings ^^

**Review by pioardi — APPROVED** [review]
Looks really good now that was another important thing that we wanted to fix so thank you much. 
Please wait for @jerome-benoit feedbacks before to merge

File: src/pools/dynamic-pool.ts
Comment (jerome-benoit) [inline src/pools/dynamic-pool.ts:20]: Are you sure to still thinking a placeholder for types definition is not necessary? :)

File: src/pools/cluster/dynamic.ts
Comment (jerome-benoit) [inline src/pools/cluster/dynamic.ts:58]: I'm not very comfortable with a method returning another method call ... 
I would expect here to be able to register the callback that will be used to choose a worker: 

+ fixed -> register the function roundRobinSelection
+ dynamic -> register the function dynamicallyChooseWorker

File: src/pools/dynamic-pool.ts
Comment (Shinigami92) [reply src/pools/dynamic-pool.ts:20]: I'm not sure if I understood you correctly :confused:  
Sorry if something was lost in translation

I created this interface so Dynamic Pools are forced to implement the `max` property
The member functions still cannot be defined by an interface, cause they are `protected`
Functions forced by an interface must be `public` by design and we don't want that specifically by our design.

Basically I write TypeScript types/interfaces to represent the structure and real estimates of the project. They are like contract definitions to shape the API and/or function in- and out-puts.

File: src/pools/cluster/dynamic.ts
Comment (pioardi) [reply src/pools/cluster/dynamic.ts:58]: @jerome-benoit  yes this is something like my proposal, having a property ( i.e. isDynamic ) and based on that property take actions, so we can even move stuff to the abstract class

Comment (Shinigami92) [reply src/pools/cluster/dynamic.ts:58]: Problem about this is that the dynamic choose worker function does so many (3) things that rely on stuff within the DynamicPool class itself that it cannot be just a function that can be swapped.

The fixed pool already do the round robin basically because it uses the default that is defined by/in AbstractPool

I think I know what your thoughts are. Just doing like

```ts
class DynamicCPool ... {
 chooseStrategy = (args) => Worker
}

class FixedCPool ... {
 chooseStrategy = (args) => Worker
}
```

but you are loosing the whole type-safety I think :thinking:

**Comment by pioardi** [comment]
Approval removed to look into some other things about those changed :) I am sure that also jerome can provide some draft starting from this branch ( I am in review mode :D )

File: src/pools/cluster/dynamic.ts
Comment (jerome-benoit) [reply src/pools/cluster/dynamic.ts:58]: > ```ts
> class DynamicCPool ... {
>  chooseStrategy = (args) => Worker
> }
> 
> class FixedCPool ... {
>  chooseStrategy = (args) => Worker
> }
> ```

I think that fine to use `...args: any[]`. 
You can't have a one size fits all here regarding types for algo that requires very different inputs.

Comment (jerome-benoit) [reply src/pools/cluster/dynamic.ts:58]: Also, if you can define the function to use for worker choice and register it, you can then use arrow function and they will automatically be binded to this.

File: src/pools/dynamic-pool.ts
Comment (jerome-benoit) [reply src/pools/dynamic-pool.ts:20]: At some point, usually code starts to gather of all them in the same place. It's just nitpicking on code modularity.

**Comment by pioardi** [comment]
So is not clear to me how we want to move on about this topic @jerome-benoit  @Shinigami92 .
My suggestion is to let @jerome-benoit  make changes on this branch or another and then do a final review.

Please let me know, in the meanwhile I would try to fix our EPIPE error that I hate to see

**Comment by Shinigami92** [comment]
@jerome-benoit could provide a proposal as a new PR on top of this PR, so we can see a better review/comparison between the current solution and his solution

---

IMO this PR currently solves its purpose of providing code that removes duplicated code
Maybe it is not really nice and can be improved, but in the current stage it solves the issue.
And it is doing so without breaking any design choices like protection of member function.

**Comment by pioardi** [comment]
> @jerome-benoit could provide a proposal as a new PR on top of this PR, so we can see a better review/comparison between the current solution and his solution
> 
> IMO this PR currently solves its purpose of providing code that removes duplicated code
> Maybe it is not really nice and can be improved, but in the current stage it solves the issue.
> And it is doing so without breaking any design choices like protection of member function.

Yes is true that solve the issue and is also true the can be improved.
I am not against merge those changes, just want wait  jerome to share alternatives/improvements.

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:33]: @pioardi @jerome-benoit 

We could replace this with
```ts
export function roundRobinSelection<Worker extends IWorker, Data, Response> (
  poolReference: AbstractPool<Worker, Data, Response>
): Worker {
  const chosenWorker = poolReference.workers[poolReference.nextWorkerIndex]
  poolReference.nextWorkerIndex =
    poolReference.workers.length - 1 === poolReference.nextWorkerIndex
      ? 0
      : poolReference.nextWorkerIndex + 1
  return chosenWorker
}
```

That way we can pass the pool directly and can remove the `RoundRobinSelectionResult` and so on.
The only downside of that would be that it is not unit-testable on its own and rely on the pool. But that would be okay for me.

What do you think?

**Comment by jerome-benoit** [comment]
> > @jerome-benoit could provide a proposal as a new PR on top of this PR, so we can see a better review/comparison between the current solution and his solution
> > IMO this PR currently solves its purpose of providing code that removes duplicated code
> > Maybe it is not really nice and can be improved, but in the current stage it solves the issue.
> > And it is doing so without breaking any design choices like protection of member function.
> 
> Yes is true that solve the issue and is also true the can be improved.
> I am not against merge those changes, just want wait jerome to share alternatives/improvements.

That PR have performance regression. I've explained them and how to fix them with 
+ function registration in a attribute
+ use of arrow function
previously. 

I'm very busy at the moment, so I can't tell when I will be able to fix them myself.

**Comment by Shinigami92** [comment]
> That PR have performance regression. I've explained them and how to fix them with
> 
> * function registration in a attribute
> * use of arrow function
> 
> previously.
> 
> I'm very busy at the moment, so I can't tell when I will be able to fix them myself.

I'm also currently a bit busy managing 3-4 Open Source projects and working for my employer ^^
But we "fixed" the performance regression via binding early in the constructor
So this problem was already fixed if I'm correct

But I found another possible improvement https://github.com/pioardi/poolifier/pull/176#discussion_r578344125 I could apply if you both agree that it could be better

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:33]: It's too complicated for one strategy implementation and not the same as the other strategy. The have to share a common signature to allow to add easily more later.

**Comment by jerome-benoit** [comment]
>
> I'm also currently a bit busy managing 3-4 Open Source projects and working for my employer ^^
> But we "fixed" the performance regression via binding early in the constructor
> So this problem was already fixed if I'm correct
> 

The external function to make the worker choice is still not bound to the object instance. You can use arrow function inside the class method to call them to ensure they will be bound to this.

**Comment by pioardi** [comment]
> > I'm also currently a bit busy managing 3-4 Open Source projects and working for my employer ^^
> > But we "fixed" the performance regression via binding early in the constructor
> > So this problem was already fixed if I'm correct
> 
> The external function to make the worker choice is still not bound to the object instance. You can use arrow function inside the class method to call them to ensure they will be bound to this.

We don't need a rush guys, @jerome-benoit  just  kindly contribute on this when you a little more time we will hold on for now :)

**Comment by Shinigami92** [comment]
@pioardi @jerome-benoit If jerome's draft get merged into my PR, please give me the chance to review it (maybe tomorrow) because I don't have that much time today

**Comment by pioardi** [comment]
@Shinigami92  take your time my dear friend :)

**Comment by Shinigami92** [comment]
@jerome-benoit (@pioardi) we currently cannot cover lines for e.g. `LessRecentlyUsedWorkerChoiceStrategy` because we don't export it
Therefore we cannot _require_ them directly in tests and unit-test them

Should we export them internally (not via `src/index.ts`) so we can write extra tests for these lines?

**Comment by Shinigami92** [comment]
We may need to update the changelog because we now have some member functions public that weren't public before
Also we need to add the new strategy to the changelog

**Comment by jerome-benoit** [comment]
> @jerome-benoit (@pioardi) we currently cannot cover lines for e.g. `LessRecentlyUsedWorkerChoiceStrategy` because we don't export it
> Therefore we cannot _require_ them directly in tests and unit-test them

I think pool unit tests can cover strategies without exporting more: a pool should be able to be created with three workers one with two tasks, one with one task and one with one task and ensure that the submitted task is running on the one with one task and the higher array index.

**Comment by Shinigami92** [comment]
> I think pool unit tests can cover strategies without exporting more: a pool should be able to be created with three workers one with two tasks, one with one task and one with one task and ensure that the submitted task is running on the one with one task and the higher array index.

Yeah already have this, but I don't have a pool with a worker with **zero** tasks and then the strategy selects the one with this zero tasks :thinking:

**Comment by Shinigami92** [comment]
@jerome-benoit I found out why this line is NEVER covered
It's because the `findFreeWorkerBasedOnTasks` function will be executed before the `choose` function will be called, the therefore within this it will never find one with number of tasks equals zero

@pioardi So that's why we cannot reach a higher coverage in this PR for now :confused: 

If you like, we should merge it now

**Comment by jerome-benoit** [comment]
> @jerome-benoit I found out why this line is NEVER covered
> It's because the `findFreeWorkerBasedOnTasks` function will be executed before the `choose` function will be called, the therefore within this it will never find one with number of tasks equals zero

It's only true for dynamic pool. For static, it shall be covered.

**Comment by Shinigami92** [comment]
> It's only true for dynamic pool. For static, it shall be covered.

Oh wow nice thx

**Comment by Shinigami92** [comment]
@jerome-benoit if you approve it, I can squash-merge it :slightly_smiling_face:

File: src/pools/cluster/fixed.ts
Comment (pioardi) [inline src/pools/cluster/fixed.ts:62]: Is this public for a particular reason now ?
Did we tracked this change into the change log?

**Review by pioardi — COMMENTED** [review]
The general shape is really good.
I just have one comment on a method that now is public.
I would also document the new strategy option on our doc ( we can do that later )
And coverage will be decreased to 92,8 % after this PR merge ( if you want you an also do that later )

Thank you very much for all your work on that

File: src/pools/cluster/fixed.ts
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:62]: Yes and no

We needed to open it (https://github.com/pioardi/poolifier/pull/191#issuecomment-782707329)
And no we havn't added this to the changelog yet

**Comment by Shinigami92** [comment]
Yeah let's do this later (we can clean up everything for the release in #89, that's why I have created that issue :wink:)

If we merge this now, I can better help on the benchmark PR

File: src/pools/cluster/fixed.ts
Comment (jerome-benoit) [reply src/pools/cluster/fixed.ts:62]: Dynamic pool worker choice strategy require to access some methods in the abstract class. 
Not sure it will last if we make strategies code not manipulate the pool internals, but for now it's the less ugly solution without a complete rewrite of the dynamic worker choice code.

**Review by jerome-benoit — APPROVED** [review]

**Comment by pioardi** [comment]
Sounds great, good job here guys

**Pull request merged**
